nixos/modules/virtualisation/azure-common.nix

```diff
@@ -1,67 +1,85 @@
-  services.openssh.enable = true;
-  services.openssh.settings.PermitRootLogin = "prohibit-password";
-  services.openssh.settings.ClientAliveInterval = 180;
-
-  # Force getting the hostname from Azure
-  networking.hostName = mkDefault "";
-
-  # Always include cryptsetup so that NixOps can use it.
-  # sg_scan is needed to finalize disk removal on older kernels
-  environment.systemPackages = [ pkgs.cryptsetup pkgs.sg3_utils ];
-
-  networking.usePredictableInterfaceNames = false;
-
-  services.udev.extraRules = ''
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:0", ATTR{removable}=="0", SYMLINK+="disk/by-lun/0",
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:1", ATTR{removable}=="0", SYMLINK+="disk/by-lun/1",
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:2", ATTR{removable}=="0", SYMLINK+="disk/by-lun/2"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:3", ATTR{removable}=="0", SYMLINK+="disk/by-lun/3"
-
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:4", ATTR{removable}=="0", SYMLINK+="disk/by-lun/4"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:5", ATTR{removable}=="0", SYMLINK+="disk/by-lun/5"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:6", ATTR{removable}=="0", SYMLINK+="disk/by-lun/6"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:7", ATTR{removable}=="0", SYMLINK+="disk/by-lun/7"
-
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:8", ATTR{removable}=="0", SYMLINK+="disk/by-lun/8"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:9", ATTR{removable}=="0", SYMLINK+="disk/by-lun/9"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:10", ATTR{removable}=="0", SYMLINK+="disk/by-lun/10"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:11", ATTR{removable}=="0", SYMLINK+="disk/by-lun/11"
-
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:12", ATTR{removable}=="0", SYMLINK+="disk/by-lun/12"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:13", ATTR{removable}=="0", SYMLINK+="disk/by-lun/13"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:14", ATTR{removable}=="0", SYMLINK+="disk/by-lun/14"
-    ENV{DEVTYPE}=="disk", KERNEL!="sda" SUBSYSTEM=="block", SUBSYSTEMS=="scsi", KERNELS=="?:0:0:15", ATTR{removable}=="0", SYMLINK+="disk/by-lun/15"
-
-  '';
-
+  imports = [
+    ../profiles/headless.nix
```

Comment:
@codgician You write in the PR description

> Remove reference of `headless` profile to make the console input work (and being able to select boot entries).

But then you'r eadding `profiles/headless.nix` back here, so that conflicts.

Does the current config actually allow you to select GRUB boot menus?

Also, what does `headless.nix` have to do with GRUB entries?

There are no GRUB related options being set here:

https://github.com/codgician/nixpkgs/blob/94683f9e55d47996b7403791863143f7afad8a51/nixos/modules/profiles/headless.nix

The only thing that's disabled is the serial console of _Linux_ -- which is also bad breaks the Azure Serial Console working on NixOS when Linux is booted, but should not affect GRUB menu selection.

Can you elaborate?

Another question:

How does this work at all when `v2` is used, given that  [`boot.loader.grub.efiSupport`](https://search.nixos.org/options?channel=24.05&show=boot.loader.grub.efiSupport&from=0&size=50&sort=relevance&type=packages&query=grub+efi) defaults to `false`?

---

I'm currently fixing it like this, for GRUB on V2 (which is EFI):

```nix
    boot.loader.systemd-boot.enable = false;
    boot.loader.grub = {
      enable = true;
      efiSupport = true;
      # Override needed because nixpkgs's `azure-common.nix` sets `device = "/dev/sda"`,
      # which renders into a mergeable list option that would include that AND `nodev`.
      devices = lib.mkOverride 99 [ "nodev" ];
      efiInstallAsRemovable = true; # see note [hetzner-canTouchEfiVariables] for an explanation
      # Enable Azure Serial Console to work with GRUB.
      # See https://learn.microsoft.com/en-us/troubleshoot/azure/virtual-machines/linux/serial-console-grub-proactive-configuration
      # Note previous serial output may not be cleared by GRUB, making
      # it hard to read the menu until you navigate with the arrow keys.
      extraConfig = ''
        serial --unit=0 --speed=115200 --word=8 --parity=no --stop=1
        terminal_input --append serial
        terminal_output --append serial
      '';
    };
    boot.loader.efi.canTouchEfiVariables = false;
```

Note the `extraConfig` to make the GRUB2 serial output work.

Screenshot of it working in Azure:

![Screenshot from 2024-11-26 00-43-51](https://github.com/user-attachments/assets/3a722ade-d0f6-4359-9091-900ef1e3c591)


And I'm re-enabling the serial output of Linux like this:

```nix
    # Explicitly enable serial-getty on ttyS0, otherwise a login prompt will only
    # be shown on the first serial device (hvc0 in our case).
    # See: http://0pointer.de/blog/projects/serial-console.html (section Serial Terminals)
    systemd.services."serial-getty@ttyS0" = {
      enable = true;
      wantedBy = [ "getty.target" ]; # to start at boot
      serviceConfig.Restart = "always"; # restart when session is closed
    };
  };
```

Reply:
For #1, you are right and thanks for the nice catch. I might have overlooked when creating this PR.

For #2, I've been using `systemd-boot` in my test flake. I did not elaborate much on the grub + UEFI scenario so I kept the pre-existing defaults. Thanks for your report and I will try to fix that for v2 generation.

I will create a new PR to address them asap.

Reply:
I have published PR #359365 after testing via flake [codgician/azure-aarch64-nixos](https://github.com/codgician/azure-aarch64-nixos). Now grub2 console output should work for both Gen 1 and Gen 2 VM.